code/data_structures/src/splay_tree/splay.cpp

```diff
@@ -0,0 +1,283 @@
+#include <iostream>
+#include <stdlib.h>
+
+struct Node
+{
+    int value;
+    Node *left;
+    Node *right;
+    Node *parent;
+};
+Node *g_root;
+
+void inOrder(Node *cur, std::string &s)
+{
+    if(cur==NULL)
+        return;
+    if(cur->left)
+        inOrder(cur->left, s);
+    s += std::to_string(cur->value) + " ";
+    if(cur->right)
+        inOrder(cur->right, s);
+}
+void rightRotate(Node *z)
+{
+    Node *l = z->left;
+    Node *c = l->right;
+    Node *p = z->parent;
+    if (p != NULL)
+    {
+        if (p->left == z)
+            p->left = l;
+        else 
+            p->right = l;
+    }
+    l->parent = p;
+    l->right = z;
+    z->parent = l;
+    z->left = c;
+    if (c != NULL)
+        c->parent = z;
+}
+
+void leftRotate(Node *z)
+{
+    Node *r = z->right;
+    Node *c = r->left;
+    Node *p = z->parent;
+    if (p != NULL)
+    {
+        if (p->left == z)
+            p->left = r;
+        else 
+            p->right = r;
+    }
+    r->parent = p;
+    r->left = z;
+    z->parent = r;
+    z->right = c;
+    if (c != NULL)
+        c->parent = z;
+}
+
+void splay(Node *z)
+{
+    if (z == NULL)
+        return;
+    while (true)
+    {
+        Node *par = z->parent;
+        if (par == NULL)
+        {
+            // z is the root
+            break;
+        }
+        Node *gPar = par->parent;
+        if (gPar == NULL && par->left == z)
+        {
+            // zig
+            rightRotate(par);
+        }
+        else if (gPar == NULL && par->right == z)
+        {
+            // zag
+            leftRotate(par);
+        }
+        else if (gPar->left == par && par->left == z)
+        {
+            // zig-zig
+            rightRotate(gPar);
+            rightRotate(par);
+        }
+        else if (gPar->right == par && par->right == z)
+        {
+            // zag-zag
+            leftRotate(gPar);
+            leftRotate(par);
+        }
+        else if (gPar->left == par && par->right == z)
+        {
+            // zig-zag
+            leftRotate(par);
+            rightRotate(gPar);
+        }
+        else if(gPar->right == par && par -> left ==z)
+        {
+            // zag-zig
+            rightRotate(par);
+            leftRotate(gPar);
+        }
+    }
+    g_root = z;
+}
+
+void insert(int val)
+{
+    Node *t = new Node();
```

Comment:
Instead of initializing all of these values yourself, a constructor would make this easier.